### PR TITLE
[com_contenthistory] - fix for not delete keep forever items

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #20365 .

### Summary of Changes
added the missed check

### Testing Instructions
see #20365

### Expected result
not able to delete a keep forever marked content history item

### Actual result
able

### CODE REVIEW

**Comment by brianteeman** [comment]
If you select 5 items - one of which is flagged keep forever - then it prevents you from deleting them all. I expected it to delete 4 items

**Comment by alikon** [comment]
should be better now, it deletes only what you can delete

File: administrator/components/com_contenthistory/models/history.php [hunk 1]
```diff
@@ -118,8 +118,15 @@ public function delete(&$pks)
 		// Iterate the items to delete each one.
 		foreach ($pks as $i => $pk)
 		{
+
```
Comment (Quy) [inline administrator/components/com_contenthistory/models/history.php]: Delete blank line

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 21d69332c7616c589076cedc4159d29331b1aa0f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20430">issues.joomla.org/tracker/joomla-cms/20430</a>.</sub>

**Comment by ggppdk** [comment]
The keep for ever feature is not "prevent user deletion" thing

It is keep a version when it "ages" due to new versions created ...
aka prevent **automatic** deletion of it

That is why the "... forever" thing 
otherwise it would have been called something else "Not deletable" or something

- please see the description of the feature in the language strings

COM_CONTENTHISTORY_BUTTON_KEEP_TOGGLE_OFF="Select to allow this version to be deleted automatically according to the delete schedule."
COM_CONTENTHISTORY_BUTTON_KEEP_TOGGLE_ON="Select to prevent this version being deleted automatically."

There is nothing to be fixed

if you do this change
then you will also need to update language strings and any documentation written for the feature
and also rename the feature

**Comment by alikon** [comment]
this is outside of the scope of this silly pr, as for my understanding of this feature...
...but i'm open to different opinions...

**Comment by ggppdk** [comment]
I would have posted to the original issue #20365, but that is now closed

it is not the PR that is wrong, it is the original issue that was wrong

**Comment by alikon** [comment]
can you show me where/when  this **automatic** deletion happens  :question:

**Comment by ggppdk** [comment]
When a new version is stored by store() method of ContentHistoryHelper

https://github.com/joomla/joomla-cms/blob/staging/libraries/src/Helper/ContentHistoryHelper.php#L147-L163

there is a cleanup of old versions to enforce the limit of maximum versions kept

and it calls 
`public function deleteOldVersions($maxVersions)`
of ContentHistory 's Table class

https://github.com/joomla/joomla-cms/blob/staging/libraries/src/Table/ContentHistory.php#L179-L200

- Inside it the "keep_forever" is checked and the versions having it are excluded from deletion

**Comment by peteruoi** [comment]
If you keep 5 versions and the first version is marked  keep forever when the 6th version will come it will not delete the first one but the second one and the result will be still 5 versions stored.

**Comment by alikon** [comment]
ok sorry this  **keep_forever**  feature, then is not clear to me

**Comment by Quy** [comment]
Closing as expected per `Keep Forever` tooltips:
`Select to allow this version to be deleted automatically according to the delete schedule.`
`Select to prevent this version being deleted automatically.`<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20430">issues.joomla.org/tracker/joomla-cms/20430</a>.</sub>

**Comment by alikon** [comment]
are you so sure ? 
then fine to me

**Comment by Quy** [comment]
@alikon I thought that too until reading the documentation further:

> Maximum Versions. The maximum number of versions to store for an item. If an item is saved and the maximum number of versions has been reached, the oldest version will be deleted automatically. If set to 0, then versions will never be deleted automatically. Also, specific versions may be flagged as "Keep Forever" and will not be deleted automatically. Note that versions may be deleted manually using the Delete button in the Version History screen.

> Keep On/Off. This button allows you to toggle on or off the Keep Forever feature for a version. Normally, the oldest version of an item will be deleted automatically when the maximum number of versions (set in the Options for the component) has been exceeded. If you set the Keep Forever property for a version, it will never be automatically deleted.

**Comment by brianteeman** [comment]
The current way keep forever works is that it wont be automatically deleted based on the number of revisions stored. However, and as originally reported, it seems perfectly reasonable to me that you also can not manually delete something marked as keep forever

**Comment by carlitorweb** [comment]
IMO have sense this PR. Mark a version as "Keep Forever" is a way to prevent someone from mistakenly deleting a version that can not be deleted.

**Comment by Quy** [comment]
@brianteeman Please decide. I tried to reopen it and got this error:

```
422 RuntimeException
Invalid response from GitHub
```

**Comment by carlitorweb** [comment]
> I tried to reopen it and got this error

@Quy Maybe because the branch was deleted?

**Comment by brianteeman** [comment]
I can't reopen it as the branch is deleted

**Comment by Quy** [comment]
@alikon Please restore the deleted branch. Sorry I made a hasty decision. Thank you!

**Comment by ggppdk** [comment]
This PR was created to fix a bug
and as detailed explained,
there was no bug to be fixed in the first place

but since majority here, thinks to change existing design
you will also need to
- update language strings and maybe rename the feature
- update documention
- add some message why the version was not deleted, 

finally people using this so far will have a small surprise when trying to cleanup their keep forever versions because they want to mark some other as keep forever

**Comment by ggppdk** [comment]
And besides what i mentioned above,
you are not changing the behavior in com_content only

you are changing the behavior in all core and 3rd party extensions that use versions
maybe a confirmation when clicking the delete button would be better: 

Some of the versions selected for deletion are marked as keep-forever, are you sure to delete them ?

**Comment by brianteeman** [comment]
@ggppdk I dont agree - keep forever should mean exactly that. No need to change the text at all

**Comment by ggppdk** [comment]
well about **name**, you are propably right, no need to change it
i said, maybe naming needs to be changed

but the description language strings and documentation need to be changed / updated
and the rest i said , do apply

**Comment by alikon** [comment]
restored
@Quy,  no problem at all :smile:

**Comment by Quy** [comment]
Reopening for further discussion.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20430">issues.joomla.org/tracker/joomla-cms/20430</a>.</sub>

File: administrator/components/com_contenthistory/models/history.php [hunk 2]
```diff
@@ -120,6 +120,12 @@ public function delete(&$pks)
 		{
 			if ($table->load($pk))
 			{
+				if ($table->keep_forever === "1")
+				{
+					unset($pks[$i]);
+					continue;
```
Comment (carlitorweb) [inline administrator/components/com_contenthistory/models/history.php]: Whitespace at end of the line

**Review by carlitorweb — CHANGES_REQUESTED** [review]

**Comment by carlitorweb** [comment]
Also, I think is good change this string `COM_CONTENTHISTORY_BUTTON_DELETE_DESC` describing that these types of versions cannot be eliminated

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 21d69332c7616c589076cedc4159d29331b1aa0f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20430">issues.joomla.org/tracker/joomla-cms/20430</a>.</sub>

**Comment by ghost** [comment]
@Quy can you please Retest?

**Comment by carlitorweb** [comment]
I did the test, just as have a CS, I not put the test result, waiting for that. I will do now, no problem

**Comment by carlitorweb** [comment]
I have tested this item :white_check_mark: successfully on 21d69332c7616c589076cedc4159d29331b1aa0f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20430">issues.joomla.org/tracker/joomla-cms/20430</a>.</sub>

**Comment by ghost** [comment]
Ready to Commit after 3 successful tests.

**Comment by alikon** [comment]
sorry @carlitorweb i've losed your comment ...

**Comment by carlitorweb** [comment]
No problem @alikon

**Pull request merged**